objects/src/accounts/account_id.rs

```diff
@@ -134,336 +129,471 @@ impl FromStr for AccountStorageMode {
+
+    const VERSION_0_NUMBER: u8 = 0;
+    /// Version 0 of the [`AccountId`].
+    pub const VERSION_0: AccountVersion = AccountVersion(Self::VERSION_0_NUMBER);
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the version number.
+    pub const fn version_num(&self) -> u8 {
+        self.0
+    }
+}
+
 // ACCOUNT ID
 // ================================================================================================
 
-/// Unique identifier of an account.
+/// The identifier of an [`Account`](crate::accounts::Account).
+///
+/// # Layout
 ///
-/// Account ID consists of 1 field element (~64 bits). The most significant bits in the id are used
-/// to encode the account' storage and type.
+/// An `AccountId` consists of two field elements and is layed out as follows:
 ///
-/// The top two bits are used to encode the storage type. The values [PRIVATE] and [PUBLIC]
-/// encode the account's storage type. The next two bits encode the account type. The values
-/// [FUNGIBLE_FAUCET], [NON_FUNGIBLE_FAUCET], [REGULAR_ACCOUNT_IMMUTABLE_CODE], and
-/// [REGULAR_ACCOUNT_UPDATABLE_CODE] encode the account's type.
+/// ```text
+/// 1st felt: [zero bit | random (55 bits) | storage mode (2 bits) | type (2 bits) | version (4 bits)]
+/// 2nd felt: [anchor_epoch (16 bits) | random (40 bits) | 8 zero bits]
+/// ```
+///
+/// # Generation
+///
+/// An `AccountId` is a commitment to a user-generated seed, the code and storage of an account and
+/// to a certain hash of an epoch block of the blockchain. An id is generated by picking an epoch
+/// block as an anchor - which is why it is also referred to as the anchor block - and creating the
+/// account's initial storage and code. Then a random seed is picked and the hash of (SEED,
+/// CODE_COMMITMENT, STORAGE_COMMITMENT, ANCHOR_BLOCK_HASH) is computed. If the hash's first element
+/// has the desired storage mode, account type, version and the high bit set to zero, the
+/// computation part of the ID generation is done. If not, another random seed is picked and the
+/// process is repeated.
+///
+/// The second felt of the ID is the second element of the hash function. Its upper 16 bits are
+/// overwritten with the epoch in which the ID is anchored and the lower 8 bits are zeroed. Thus,
+/// the first felt of the ID must derive exactly from the hash, while only part of the second felt
+/// is derived from the hash.
+///
+/// # Constraints
+///
+/// Constructors will return an error if:
+///
+/// - There are fewer than [`AccountId::MIN_ACCOUNT_ONES`] in the first felt.
+/// - The first felt's most significant bit is not zero.
+/// - The first felt contains account ID metadata (storage mode, type or version) that does not
+///   match any of the known values.
+/// - The anchor epoch in the second felt is equal to [`u16::MAX`].
+/// - The lower 8 bits of the second felt are not zero.
+///
+/// # Design Rationale
```

Comment:
Should this section be part of mod-level docs?

Reply:
I like this idea, but since we don't have mod-level docs right now it would feel like lacking context if I added only the rationale there. And it would fragment the description of the layout and the rationale, so this might need to be done as a more extensive account Rust docs refactor, so I'd leave it as-is for now if that's fine.